xdsl/dialects/riscv.py

```diff
@@ -157,13 +164,107 @@ def print_parameter(self, printer: Printer) -> None:
 
 
 class RISCVOp(Operation, ABC):
-    pass
+    """
+    Base class for operations that can be a part of RISC-V assembly printing.
+    """
+
+    @abstractmethod
+    def assembly_line(self) -> str | None:
+        raise NotImplementedError()
+
+
+_AssemblyInstructionArg: TypeAlias = (
+    AnyIntegerAttr | LabelAttr | SSAValue | RegisterType | str | None
+)
+
+
+class RISCVInstruction(RISCVOp):
+    """
+    Base class for operations that can be a part of RISC-V assembly printing. Must
+    represent an instruction in the RISC-V instruction set, and have the following format:
+
+    name arg0, arg1, arg2           # comment
+
+    The name of the operation will be used as the RISC-V assembly instruction name.
+    """
+
+    comment: OptOpAttr[StringAttr]
+    """
+    An optional comment that will be printed along with the instruction.
+    """
+
+    @abstractmethod
+    def assembly_line_args(self) -> tuple[_AssemblyInstructionArg, ...]:
+        """
+        The arguments to the instruction, in the order they should be printed in the
+        assembly.
+        """
+        raise NotImplementedError()
+
+    def assembly_line(self) -> str | None:
+        # default assembly code generator
+        assert self.name.startswith("riscv.")
+        instruction_name = self.name[6:]
```

Comment:
Do we want to assert that the op starts with `riscv.`? I feel that maybe another dialect may want to have ops in it that are also risc-v printable, and it would feel wrong to have operations in a dialect with multiple prefixes (e.g. the `snrt` dialect would have `snrt` and `riscv.` operations).

Maybe we just want to remove the first segmetn of the op name?

Also, I feel we should document this contract somewhere.

Reply:
good point, I'll change this